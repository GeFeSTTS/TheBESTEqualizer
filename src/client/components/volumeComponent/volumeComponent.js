import React, { Component, Fragment } from 'react';
import 'react-rangeslider/lib/index.css';
import './volumeComponent.css';
import Slider from 'react-rangeslider';
import { connect } from 'react-redux';
import { volumeIcon } from '../../assets/icons/icons';

export class VolumeComponent extends Component {
  state = {
    volumeValueSound: 0.25,
    sound: null,
  };

  static getDerivedStateFromProps(props, state) {
    const { audioData: { sound, voice } } = props;
    const { volumeValueSound } = state;
    if (sound && sound.playing) {
      sound.volume = volumeValueSound;
      return { sound };
    }
    if (voice && voice.playing) {
      voice.volume = volumeValueSound;
      return { sound: voice };
    }
    return null;
  }

  changeTracksOrVoiceVolume = () => {
    const {
      volumeValueSound, sound,
    } = this.state;
    if (sound) {
      sound.volume = volumeValueSound;
    }
  }

  changeVolume = (volumeValueSound) => {
    this.setState({
      volumeValueSound: parseFloat(volumeValueSound),
    }, () => this.changeTracksOrVoiceVolume());
  }

  formatValue = (volumeValueSound) => {
    const numberForFormatingVolume = 200;
    return Math.round(volumeValueSound * numberForFormatingVolume);
  }

  render() {
    const { volumeValueSound } = this.state;
    const minSliderVolume = 0;
    const maxSliderVolume = 0.5;
    const stepSliderVolume = 0.001;
    return (
      <div className="SwitcherContainer">
        <Fragment>
          <span>
            {volumeIcon}
            {' '}
Volume
          </span>
          <Slider
            className="SwitcherContainer--slider"
            value={volumeValueSound}
            format={this.formatValue}
            min={minSliderVolume}
            max={maxSliderVolume}
            step={stepSliderVolume}
            onChange={this.changeVolume}
          />
        </Fragment>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  audioData: state.audioData,
});

export default connect(mapStateToProps)(VolumeComponent);
