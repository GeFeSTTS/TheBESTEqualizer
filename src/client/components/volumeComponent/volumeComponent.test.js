import React from 'react';
import { shallow } from 'enzyme';
import { VolumeComponent } from './volumeComponent';

describe('TEST VOLUME SLIDER', () => {
  let VolumeComponentWrapper;

  const props = {
    min: 0,
    max: 1,
    step: 0.001,
    onChange: () => {},
    value: 0,
    audioData: { sound: {}, voice: {} },
  };

  beforeEach(() => {
    VolumeComponentWrapper = shallow(<VolumeComponent {...props} />);
  });

  it('is render component', () => {
    expect(VolumeComponentWrapper.find('.SwitcherContainer')).toHaveLength(1);
  });

  it('render component correctly', () => {
    expect(VolumeComponentWrapper).toMatchSnapshot();
  });

  it('should call function after change', () => {
    const instance = VolumeComponentWrapper.instance();
    const spySwitching = jest.spyOn(instance, 'changeVolume');

    VolumeComponentWrapper.setState({ volumeValueSound: 0.8 });
    VolumeComponentWrapper.instance().forceUpdate();
    VolumeComponentWrapper.find('Slider').simulate('change', 1);

    expect(spySwitching).toHaveBeenCalled();
  });

  it('should we return sound as track in getDerivedStateFromProps', () => {
    VolumeComponentWrapper.setProps({ audioData: { sound: { playing: true }, voice: null } });

    expect(VolumeComponentWrapper).toHaveLength(1);
  });

  it('should we return sound as voice in getDerivedStateFromProps', () => {
    VolumeComponentWrapper.setProps({ audioData: { sound: null, voice: { playing: true } } });

    expect(VolumeComponentWrapper).toHaveLength(1);
  });
});
