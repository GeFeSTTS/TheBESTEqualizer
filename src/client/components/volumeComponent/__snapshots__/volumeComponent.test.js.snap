// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST VOLUME SLIDER render component correctly 1`] = `ShallowWrapper {}`;
